packages/payment-detection/src/meta-payment-detector.ts

```diff
@@ -0,0 +1,190 @@
+    }
+    return this.extension.createCreationAction({
+      ...paymentNetworkCreationParameters,
+    });
+  }
+
+  /**
+   * Creates the extensions data to apply an action on a sub pn
+   *
+   * @param Parameters to add refund information
+   * @returns The extensionData object
+   */
+  public createExtensionsDataForApplyActionOnPn(
+    parameters: ExtensionTypes.PnMeta.IApplyActionToPn,
+  ): ExtensionTypes.IAction {
+    return this.extension.createApplyActionToPn({
+      pnIdentifier: parameters.pnIdentifier,
+      action: parameters.action,
+      parameters: parameters.parameters,
+    });
+  }
+
+  /**
+   * To retrieve all events, iterate over the sub payment networks and aggregate their balances
+   */
+  protected async getEvents(
+    request: RequestLogicTypes.IRequest,
+  ): Promise<
+    PaymentTypes.AllNetworkEvents<
+      | PaymentTypes.ConversionPaymentNetworkEventParameters
+      | PaymentTypes.IDeclarativePaymentEventParameters
+    >
+  > {
+    const paymentExtension = this.getPaymentExtension(request);
+    const events: PaymentTypes.IBalanceWithEvents<PaymentTypes.GenericEventParameters>[] = [];
+    const feeBalances: PaymentTypes.IBalanceWithEvents<PaymentTypes.GenericEventParameters>[] = [];
+    this.checkRequiredParameter(paymentExtension.values.salt, 'salt');
+
+    for (const value of Object.values(
+      paymentExtension.values as Record<string, ExtensionTypes.IState<any>>,
+    )) {
+      if (supportedPns.includes(value.id as ExtensionTypes.PAYMENT_NETWORK_ID)) {
+        const detectorClass = detectorMap[value.id as keyof typeof detectorMap];
+        const extensionKey = advancedLogicMap[value.id as keyof typeof advancedLogicMap];
+        const extension =
+          this.advancedLogic.extensions[extensionKey as keyof typeof this.advancedLogic.extensions];
+
+        if (!detectorClass || !extension) {
+          throw new Error(`the payment network id: ${value.id} is not supported`);
+        }
+
+        const detector = new detectorClass({
+          advancedLogic: this.advancedLogic,
+          paymentNetworkId: value.id as ExtensionTypes.PAYMENT_NETWORK_ID,
+          extension,
+          currencyManager: this.currencyManager,
+          ...this.options,
+        });
+
+        const partialRequest = deepCopy(request);
+        partialRequest.extensions = {
+          [value.id]: value,
+        };
+        partialRequest.extensionsData = [value];
+
+        events.push(await detector.getBalance(partialRequest));
+        const feeBalance = partialRequest.extensions[value.id].values.feeBalance;
+        if (feeBalance) {
+          feeBalances.push(feeBalance);
+        }
+      }
+    }
+    const declaredEvents = this.getDeclarativeEvents(request);
+    const allPaymentEvents = [...declaredEvents, ...events.map((event) => event.events).flat()];
+
+    // FIXME: should be at the same level as balance
+    const values: any = this.getPaymentExtension(request).values;
+    values.feeBalance = {
```

Comment:
Have you wondered about the consequences of overriding the PN values instead of amending them? I think it makes but I don't have my head enough into it to be sure.

Reply:
We do the exact same thing for the `any-to-erc20` payment detection. The `FIXME` comes from there.